### DISCUSSION TITLE
discriminator support?

### DISCUSSION THREAD

**Discussion by maxgurewitz** [sinclairzx81/typebox#828] [General]
Hello! First want to say a huge thanks for this project! Probably my favorite typescript library.

Wanted to ask, does typebox support currently, or plan to support the new `discriminator` keyword?

https://ajv.js.org/json-schema.html#discriminator

Thanks!

**Comment by sinclairzx81** [comment]
@maxgurewitz Hiya,

> Wanted to ask, does typebox support currently, or plan to support the new discriminator keyword?

TypeBox only officially supports keywords defined in the Json Schema specification, so the `discriminator` keyword would be out of scope (however the oneOf schema representation may be included one day). This said, if you are using Ajv, you can construct most schema representations and generate appropriate static types using Unsafe. This is generally the recommended way to approach generating schematics not provided by TypeBox by default.

The following implements the example schema on the Ajv documentation website. 

```typescript
import { Type, Static, TUnion, TSchema, SchemaOptions } from '@sinclair/typebox'

export function Enum<T extends string[]>(values: [...T], options: SchemaOptions = {}) {
    return Type.Unsafe<T[number]>({ ...options, enum: values })
}
export function OneOf<T extends TSchema[]>(oneOf: [...T], options: SchemaOptions = {}) {
   return Type.Unsafe<Static<TUnion<T>>>({ ...options, oneOf })
}
export const T = OneOf([
  Type.Object({
    foo: Type.Literal('x'),
    a: Type.String()
  }),
  Type.Object({
    foo: Enum(['y', 'z']),
    b: Type.String()
  })
], {
    type: 'object',
    discriminator: {propertyName: "foo"},
    required: ['foo']
})

// https://ajv.js.org/json-schema.html#discriminator
// {
//   type: "object",
//   discriminator: {propertyName: "foo"},
//   required: ["foo"],
//   oneOf: [
//     {
//       properties: {
//         foo: {const: "x"},
//         a: {type: "string"},
//       },
//       required: ["a"],
//     },
//     {
//       properties: {
//         foo: {enum: ["y", "z"]},
//         b: {type: "string"},
//       },
//       required: ["b"],
//     },
//   ],
// }
```
There may be other ways you can implement the above function (factoring the `type: object`, `descriminator` and `required` (which are passed as options in the above function)) but the above should provide a good reference point.

Hope this helps!
S

**Comment by maxgurewitz** [comment]
Thank you!
